src/metabase/util/malli/fn.clj

```diff
@@ -300,33 +300,11 @@
 ;; ------------------------------ Skipping Namespace Enforcement in prod ------------------------------
 
 (defn instrument-ns?
-  "Used to track namespaces to not enforce malli schemas on with `mu.fn/fn`."
+  "Returns true if mu.fn/fn and mu/defn in a namespace should be instrumented with malli schema validation."
   [namespace]
-  (let [lib-and-middleware [#"^metabase\.lib\..*"
-                            #"^metabase\.query-processor\.middleware\..*"
-                            #"^metabase\.sync.*"
-                            #"^metabase\.upload.*"]
-        matches?           (core/fn [namespace regexes]
-                             (let [n (-> namespace ns-name str)]
-                               (some #(re-matches % n) regexes)))
-        ;; empty but placeholder for any namespaces we want to never instrument (in prod)
-        ad-hoc             #{}
-        m                  (meta namespace)]
-    (cond (:instrument/always m)                  true
-          (:instrument/never m)                   false
-          (matches? namespace lib-and-middleware) false
-          (contains? ad-hoc (ns-name namespace))  false
-          :else                                   true)))
-
-(def ^:private ^:dynamic *skip-ns-decision-fn*
```

Comment:
I decided to replace the `*skip-ns-decision-fn*`, with `instrument-ns?`. 

Some differences:
- it returns true instead of false and vice versa. I think it is more natural to have a "do-something?" function return true when something should be done than having a "skip-something?" function return true when to not do something
- it doesn't log when instrumentation is skipped.
- it replaces the logic of `*skip-ns-decision-fn*` and the old `instrument-ns?` in one function

Reply:
So much better 😍

Reply:
The one case I can think of for `binding` over `with-redefs` is that we're moving to running more tests in parallel. We're stubbing out this method in multiple tests so they'd risk flaking. This gives me a cheeky idea for a test helper that takes the intrinsic locks on the vars before redefining them, perhaps something we could use to whittle away at more dynamic-for-tests-only things.

Reply:
That's a good idea. We use `with-redefs` in a lot of places so such a test helper would be helpful. 

I found a blog post describing exactly that:
https://medium.com/helpshift-engineering/a-study-in-parallelising-tests-b5253817beae#:~:text=Although%20it%20is%20a%20very,is%20visible%20globally%20across%20threads

At the moment the tests in this PR are not particularly costly so I think it makes sense to implement the thread-safe `with-redefs` and switch tests that use it to run in parallel as a bulk PR.

Reply:
Interesting approach, much more sophisticated than what I was thinking but nicer too. Adds a bit more overhead but don't think we'll feel it.

+1 on leaving parallelism for another sweep